src/kernel/include/shared/cos_types.h

```diff
@@ -394,11 +394,75 @@ typedef int                thdclosure_index_t;
  * This type provides an abstraction for hardware-specific 
  * protection domain identifiers (ASID, MPK) that are 
  * interpreted at the hardware-abstraction-layer level. 
- * It is intended for interfaces for hardware-aware user-level 
- * code such as namespace managers. There might be a better 
- * way to do this but it cleans up the kernel interface. 
+ * Only hardware-specific code should touch the type values
  */
-typedef unsigned long prot_domain_t;
+#if defined(__x86_64__)
+	typedef struct {
+		u16_t mpk_key;
+		u16_t asid;
+	} __attribute__((__packed__)) prot_domain_t;
```

Comment:
BTW, it is also perfectly fine just to have `prot_domain_t ` be a `u32_t`, and to use your functions to use bitops to extract the values. Either is fine, in general. Since you have this, it seems fine.